### ISSUE TITLE
bug: identity of objects in typed lists not preserved

### ISSUE THREAD

**Issue by shaunc** [numba/numba#9285]
## Reporting a bug

When adding an object to a list, I expect to get the same object back when I reference it in the list. However, numba type list seems to return a copy of the object instead.

```py
import numpy as np
import numba as nb  # type: ignore

@nb.experimental.jitclass([('x', nb.int32)])  # type: ignore
class Foo:
    def __init__(self, x: np.int32):
        self.x = x

    def __eq__(self, __value: object) -> bool:
        return isinstance(__value, Foo) and self.x == __value.x

FooType = Foo.class_type.instance_type  # type: ignore

foo = Foo(np.int32(1))

foo_list = [foo]
print(foo_list[0] is foo)  # True
foo_typed_list: list[Foo] = nb.typed.List.empty_list(FooType)  
foo_typed_list.append(foo)
print(foo_typed_list[0] == foo)  # True
print(foo_typed_list[0] is foo)  # False??
# expected True
```

- [ ] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/main/CHANGE_LOG).
- [ ] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

**Comment by esc** [comment]
@shaunc yes, I can reproduce it and it is a bit mysterious. Maybe you can try with a different refcounted object, like a Numpy array?

**Comment by shaunc** [comment]
A numpy array works as expected:
```py
a = np.array([1, 2, 3])
a_list = nb.typed.List.empty_list(nb.typeof(a))
a_list.append(a)
print(a_list[0] is a)  # True
```

**Comment by esc** [comment]
> A numpy array works as expected:
> 
> ```python
> a = np.array([1, 2, 3])
> a_list = nb.typed.List.empty_list(nb.typeof(a))
> a_list.append(a)
> print(a_list[0] is a)  # True
> ```

OK, thank you, probably a jitclass issue then.

**Comment by Azmisov** [comment]
This is just for object mode I believe. Maybe the foo_typed_list[0] is returns a new wrapper for the python interpreter which is a different wrapper from for plain foo variable. But the actual underlying objects are the same, e.g.:

```python
@jitclass
class Foo:
	x: uint32
	def __init__(self, x):
		self.x = x

	def __eq__(self, __value: object) -> bool:
		return isinstance(__value, Foo) and self.x == __value.x

FooType = as_numba_type(Foo)

@njit
def test():
	foo = Foo(1)
	lst = List.empty_list(FooType)  
	lst.append(foo)
	a = lst[0] == foo
	lst[0].x = 99
	b = foo.x == 99
	with objmode():
		print(a, b)

test()
```

**Comment by shaunc** [comment]
@Azmisov - did you try the original example? That seems to exhibit the bug without using objmode.

**Comment by Azmisov** [comment]
Your example is not wrapped in a jit function, so it isn't being compiled in noobjmode. Here's your example wrapped in jit:
```python
import numpy as np
import numba as nb  # type: ignore

@nb.experimental.jitclass([('x', nb.int32)])  # type: ignore
class Foo:
	def __init__(self, x: np.int32):
		self.x = x

	def __eq__(self, __value: object) -> bool:
		return isinstance(__value, Foo) and self.x == __value.x

FooType = Foo.class_type.instance_type  # type: ignore

@nb.jit(forceobj=True)
def test_objmode():
	foo = Foo(np.int32(1))
	foo_typed_list: list[Foo] = nb.typed.List.empty_list(FooType)  
	foo_typed_list.append(foo)
	print("objmode equal:", foo_typed_list[0] == foo) # true
	print("objmode is:", foo_typed_list[0] is foo) # false (expected it to be true)
	foo_typed_list[0].x = 999
	print("objmode modified correctly:", foo_typed_list[0].x == foo.x) # true

@nb.njit
def test_noobjmode():
	foo = Foo(np.int32(1))
	foo_typed_list: list[Foo] = nb.typed.List.empty_list(FooType)  
	foo_typed_list.append(foo)
	print("noobjmode equal:", foo_typed_list[0] == foo) # true
	# print("noobjmode is:", foo_typed_list[0] is foo) # not supported in numba
	foo_typed_list[0].x = 999
	print("noobjmode modified correctly:", foo_typed_list[0].x == foo.x) # true

test_objmode()
test_noobjmode()
```

When modified (via the `x = 999` line) for objmode or noobjmode, the change shows up in both variable references. So the underlying objects are actually the same. The only issue is `is` not working properly in objmode, which I think must mean separate wrapper objects are being returned.

In noobjmode, using `is` operator won't compile. The source code indicates it isn't supported, see: https://github.com/numba/numba/blob/db5f0a45fcccb359cba248c4767cd1caf16c4a85/numba/experimental/jitclass/overloads.py#L202-L204

Perhaps should be a feature request, or mentioned more explicitly in the docs that jitclass doesn't support `is`?

**Comment by esc** [comment]
> Hmm... but a jitclass should also work in plain python code? The use in objmode is an additional bug, but the use in plain python should work as well.

@shaunc with an emphasis on "should" -- `jitclass` is in the `experimental` subpackage -- and for a good reason! Colloquially this may be referred to as a "problem factory" (for lack of a better description) and is effectively looking for a champion. I can only speak for myself, but from the number of bug reports related to `jitclass` I have seen on the tracker over the last five years -- I would actually tend to discourage using it in production and instead encourage contributors/community members to consider submitting pull-requests to overhaul the implementation.
